### ISSUE TITLE
Visual Studio/CMake UNICODE configuration issue (also: vcpkg)

### ISSUE THREAD

**Issue by yuggieg** [log4cplus/log4cplus#417]
Hi,

I was fighting with the combination of log4cplus/Visual Studio/vcpkg the last day, only to figure out there was an issue with the UNICODE definition as used in Visual Studio but not when using CMake. Please see [here](https://github.com/microsoft/vcpkg/issues/6775).

Shouldn't the CMake targets be modified such that the downstream target has the correct UNICODE and _UNICODE defined to avoid this problem?

**Comment by wilx** [comment]
I am not entirely sure what you are asking me to change. There is already a CMake option that you can set to influence the build. The default is to build with `UNICODE` and `_UNICODE` defined on if CMake's `WIN32` is true. See <https://github.com/log4cplus/log4cplus/blob/2.0.x/CMakeLists.txt#L59>. If the microsoft/vcpkg maintainers wish to override the default they can do so.

**Comment by Neumann-A** [comment]
The auto generated target for log4cplus must contain that info/flag as a build requirement. This is automatically done for you if you would use `target_compile_definitions(<log4cplustarget> PUBLIC <flag>)` instead of `add_definitions` for flags downstream targets are also required to use.

**Comment by wilx** [comment]
I have committed the suggested change to a branch: https://github.com/wilx/log4cplus/tree/2.0.x-better-cmake-integration

Please review and test if this works for you and vcpkg builds.

**Comment by Neumann-A** [comment]
generated target looks good. 

```
set_target_properties(log4cplus::log4cplusU PROPERTIES
  INTERFACE_COMPILE_DEFINITIONS "UNICODE;_UNICODE"
  INTERFACE_INCLUDE_DIRECTORIES "${_IMPORT_PREFIX}/include"
)
```

Since I am not a user of log4cplus I will not spend time to test it, sry. But from the looks of it the definition should now get passed to downstream cmake projects.

**Comment by wilx** [comment]
I have merged the changes to 2.0.x branch.
